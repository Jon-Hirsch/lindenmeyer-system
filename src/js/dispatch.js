let dispatch;

export default function(action) {
  return dispatch(action);
}

export function initDispatch(newDispatch) {
  dispatch = newDispatch;
}
