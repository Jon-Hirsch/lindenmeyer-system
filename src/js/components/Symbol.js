import React from 'react';
import PropTypes from 'prop-types';
import dispatch from '../dispatch';

export default function Symbol({
  name,
  operation,
  amount,
  productionRules,
  id,
  firstSymbol
}) {
  const showAmount = ['move', 'rotate'].includes(operation);
  return (
    <div className="lindenmeyer-system-symbol">
      Name:{' '}
      <input
        type="text"
        className="symbol-name"
        maxLength={1}
        value={name}
        onChange={({ target: { value } }) => updateSetting(id, 'name', value)}
      />{' '}
      | Operation:{' '}
      <select
        value={operation}
        onChange={({ target: { value } }) =>
          updateSetting(id, 'operation', value)}
      >
        <option value="none">None</option>
        <option value="move">Move Foward</option>
        <option value="rotate">Rotate</option>
        <option value="startBranch">Start Branch</option>
        <option value="endBranch">End Branch</option>
      </select>{' '}
      {showAmount && (
        <input
          type="text"
          className="symbol-amount"
          value={amount}
          onChange={({ target: { value } }) =>
            updateSetting(id, 'amount', value)}
        />
      )}{' '}
      | Production Rules:{' '}
      <input
        type="text"
        className="symbol-rules"
        value={productionRules}
        onChange={({ target: { value } }) =>
          updateSetting(id, 'productionRules', value)}
      />
      {!firstSymbol && (
        <div
          className="lindenmeyer-system-button delete-symbol-button"
          onClick={() => removeSymbol(id)}
        >
          Remove Symbol
        </div>
      )}
    </div>
  );
}

function removeSymbol(id) {
  dispatch({
    type: 'REMOVE_SYMBOL',
    id
  });
}

function updateSetting(id, setting, value) {
  dispatch({
    type: 'UPDATE_SYMBOL_SETTING',
    id,
    setting,
    value
  });
}

Symbol.propTypes = {
  name: PropTypes.string.isRequired,
  operation: PropTypes.string.isRequired,
  amount: PropTypes.string.isRequired,
  productionRules: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  firstSymbol: PropTypes.bool.isRequired
};
