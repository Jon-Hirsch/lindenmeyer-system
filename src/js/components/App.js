import React from 'react';
import Symbol from './Symbol';

export default function App({ state }) {
  const {
    symbols,
    axiom,
    iterations,
    startX,
    startY,
    startAngle,
    preset
  } = state;
  return (
    <div>
      <h3>Symbols</h3>
      {symbols.map((symbol, index) => (
        <Symbol key={symbol.name + index} {...symbol} firstSymbol={!index} />
      ))}
      <div className="lindenmeyer-system-button add-symbol-button">Add Symbol</div>
      <div className="lindenmeyer-system-start-settings">
        <div>
          <h3>Axiom (Start Value)</h3>
          Axiom: <input value={axiom} />
        </div>
        <div>
          <h3>Position</h3>
          <div>
            X: <input type="text" value={startX} />
          </div>
          <div>
            Y: <input type="text" value={startY} />
          </div>
        </div>
        <div>
          <h3>Presets</h3>
          <select value={preset}>
            <option value="plant">Plant</option>
          </select>
        </div>
      </div>
      <div className="lindenmeyer-system-start-settings">
        <div>
          <h3>Number of Iterations</h3>
          Iterations:{' '}
          <input
            type="text"
            className="lindenmeyer-system-iterations"
            value={iterations}
          />
        </div>
        <div>
          <h3>Starting Angle</h3>
          Angle:{' '}
          <input
            type="text"
            className="lindenmeyer-system-angle"
            value={startAngle}
          />
        </div>
        <div />
      </div>
    </div>
  );
}
