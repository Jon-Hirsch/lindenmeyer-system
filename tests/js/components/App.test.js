import React from 'react';
import App from '../../../src/js/components/App';
import Symbol from '../../../src/js/components/Symbol';
import dispatch from '../../../src/js/dispatch';

jest.mock('../../../src/js/dispatch');

describe('App', () => {
  let shallowWrapper, state;

  function render() {
    return shallow(<App state={state} />);
  }

  beforeEach(() => {
    dispatch.mockClear();
    state = {
      symbols: [
        {
          name: 'x',
          operation: 'move',
          amount: '2',
          productionRules: '+y-x-y+',
          id: '1'
        },
        {
          name: 'y',
          operation: 'move',
          amount: '2',
          productionRules: '-x+y+x-',
          id: '2'
        },
        {
          name: '+',
          operation: 'rotate',
          amount: '60',
          productionRules: '+',
          id: '3'
        },
        {
          name: '-',
          operation: 'rotate',
          amount: '-60',
          productionRules: '-',
          id: '4'
        }
      ],
      axiom: 'x',
      iterations: '8',
      startX: '150',
      startY: '20',
      startAngle: '0',
      preset: 'triangle'
    };

    shallowWrapper = render();
  });

  test('should match snapshot', () => {
    expect(shallowWrapper.getElement()).toMatchSnapshot();
  });

  test('should render a Symbol component for each symbol', () => {
    expect(shallowWrapper.find(Symbol).length).toEqual(4);
  });

  describe('axiom field', () => {
    test('should dispatch UPDATE_SETTING on change', () => {
      shallowWrapper
        .find('.lindenmeyer-system-axiom')
        .simulate('change', { target: { value: 'newValue' } });
      expect(dispatch).toHaveBeenCalledWith({
        type: 'UPDATE_SETTING',
        setting: 'axiom',
        value: 'newValue'
      });
    });
  });

  describe('x field', () => {
    test('should dispatch UPDATE_SETTING on change', () => {
      shallowWrapper
        .find('.lindenmeyer-system-x')
        .simulate('change', { target: { value: 'newValue' } });
      expect(dispatch).toHaveBeenCalledWith({
        type: 'UPDATE_SETTING',
        setting: 'startX',
        value: 'newValue'
      });
    });
  });

  describe('y field', () => {
    test('should dispatch UPDATE_SETTING on change', () => {
      shallowWrapper
        .find('.lindenmeyer-system-y')
        .simulate('change', { target: { value: 'newValue' } });
      expect(dispatch).toHaveBeenCalledWith({
        type: 'UPDATE_SETTING',
        setting: 'startY',
        value: 'newValue'
      });
    });
  });

  describe('iterations field', () => {
    test('should dispatch UPDATE_SETTING on change', () => {
      shallowWrapper
        .find('.lindenmeyer-system-iterations')
        .simulate('change', { target: { value: 'newValue' } });
      expect(dispatch).toHaveBeenCalledWith({
        type: 'UPDATE_SETTING',
        setting: 'iterations',
        value: 'newValue'
      });
    });
  });

  describe('angle field', () => {
    test('should dispatch UPDATE_SETTING on change', () => {
      shallowWrapper
        .find('.lindenmeyer-system-angle')
        .simulate('change', { target: { value: 'newValue' } });
      expect(dispatch).toHaveBeenCalledWith({
        type: 'UPDATE_SETTING',
        setting: 'startAngle',
        value: 'newValue'
      });
    });
  });

  describe('preset dropdown', () => {
    test('should dispatch SELECT_PRESET_SHAPE on change', () => {
      shallowWrapper
        .find('.lindenmeyer-system-presets')
        .simulate('change', { target: { value: 'snowflake' } });
      expect(dispatch).toHaveBeenCalledWith({
        type: 'SELECT_PRESET_SHAPE',
        shape: 'snowflake'
      });
    });
  });

  describe('add symbol button', () => {
    test('should dispatch ADD_SYMBOL on click', () => {
      shallowWrapper.find('.add-symbol-button').simulate('click');
      expect(dispatch).toHaveBeenCalledWith({
        type: 'ADD_SYMBOL'
      });
    });
  });
});
