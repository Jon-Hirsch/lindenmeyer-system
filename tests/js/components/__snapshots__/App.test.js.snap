// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div>
  <h3>
    Symbols
  </h3>
  <_Symbol
    amount="2"
    firstSymbol={true}
    id="1"
    name="x"
    operation="move"
    productionRules="+y-x-y+"
  />
  <_Symbol
    amount="2"
    firstSymbol={false}
    id="2"
    name="y"
    operation="move"
    productionRules="-x+y+x-"
  />
  <_Symbol
    amount="60"
    firstSymbol={false}
    id="3"
    name="+"
    operation="rotate"
    productionRules="+"
  />
  <_Symbol
    amount="-60"
    firstSymbol={false}
    id="4"
    name="-"
    operation="rotate"
    productionRules="-"
  />
  <div
    className="lindenmeyer-system-button add-symbol-button"
    onClick={[Function]}
  >
    Add Symbol
  </div>
  <div
    className="lindenmeyer-system-start-settings"
  >
    <div>
      <h3>
        Axiom (Start Value)
      </h3>
      Axiom:
       
      <input
        className="lindenmeyer-system-axiom"
        onChange={[Function]}
        value="x"
      />
    </div>
    <div>
      <h3>
        Position
      </h3>
      <div>
        X:
         
        <input
          className="lindenmeyer-system-x"
          onChange={[Function]}
          type="text"
          value="150"
        />
      </div>
      <div>
        Y:
         
        <input
          className="lindenmeyer-system-y"
          onChange={[Function]}
          type="text"
          value="20"
        />
      </div>
    </div>
    <div>
      <h3>
        Presets
      </h3>
      <select
        className="lindenmeyer-system-presets"
        onChange={[Function]}
        value="triangle"
      >
        <option
          value="plant"
        >
          Plant
        </option>
        <option
          value="triangle"
        >
          Sierpinski Triangle
        </option>
        <option
          value="snowflake"
        >
          Koch Snowflake
        </option>
        <option
          value="dragonCurve"
        >
          Dragon Curve
        </option>
        <option
          value="triangleVariant"
        >
          Sierpinski Triangle Variant
        </option>
      </select>
    </div>
  </div>
  <div
    className="lindenmeyer-system-start-settings"
  >
    <div>
      <h3>
        Number of Iterations
      </h3>
      Iterations:
       
      <input
        className="lindenmeyer-system-iterations"
        onChange={[Function]}
        type="text"
        value="8"
      />
    </div>
    <div>
      <h3>
        Starting Angle
      </h3>
      Angle:
       
      <input
        className="lindenmeyer-system-angle"
        onChange={[Function]}
        type="text"
        value="0"
      />
    </div>
    <div />
  </div>
</div>
`;
